examples/graph/gqa_training.cpp

```diff
@@ -0,0 +1,567 @@
+    auto key_lt = logical_tensor(101, dt, kv_sz, layout_type::strided);
+    auto scale_lt = logical_tensor(102, dt, scale_sz, layout_type::strided);
+    auto mask_lt = logical_tensor(103, dt, score_sz, layout_type::strided);
+    auto value_lt = logical_tensor(104, dt, kv_sz, layout_type::strided);
+
+    // Create tensor objects
+    tensor ts_query(query_lt, eng);
+    tensor ts_key(key_lt, eng);
+    tensor ts_scale(scale_lt, eng);
+    tensor ts_mask(mask_lt, eng);
+    tensor ts_value(value_lt, eng);
+    tensor ts_output, ts_stats;
+
+    // Allocate and initialize data
+    std::vector<float> query_data(product(q_sz));
+    std::vector<float> key_data(product(kv_sz));
+    std::vector<float> scale_data(product(scale_sz), std::sqrt(p.head_size));
+    std::vector<float> mask_data(product(score_sz));
+    std::vector<float> value_data(product(kv_sz));
+
+    fill_random(query_data);
+    fill_random(key_data);
+    fill_random(value_data);
+    fill_mask(mask_data, static_cast<size_t>(p.seq_len));
+
+    // Write data to tensor objects
+    write_to_dnnl_tensor(query_data.data(), ts_query);
+    write_to_dnnl_tensor(key_data.data(), ts_key);
+    write_to_dnnl_tensor(scale_data.data(), ts_scale);
+    write_to_dnnl_tensor(mask_data.data(), ts_mask);
+    write_to_dnnl_tensor(value_data.data(), ts_value);
+
+    // Run forward pass
+    bench_gqa_forward(ekind, dt, strm, eng, ts_query, ts_key, ts_scale, ts_mask,
+            ts_value, ts_output, ts_stats, time_limit);
+
+    // Prepare output gradients
+    const dims doutput_sz = ts_output.get_logical_tensor().get_dims();
+    auto doutput_lt = logical_tensor(105, dt, doutput_sz, layout_type::strided);
+    tensor ts_doutput(doutput_lt, eng);
+
+    // Allocate and initialize gradients
+    std::vector<float> doutput_data(product(doutput_sz));
+    fill_random(doutput_data);
+
+    // Run backward pass
+    bench_gqa_backward(ekind, dt, strm, eng, ts_query, ts_key, ts_scale,
+            ts_mask, ts_value, ts_output, ts_stats, ts_doutput, time_limit);
+}
+
+void bad_args() {
+    std::cerr << "Usage: graph-gqa-cpp [cpu|gpu]\n"
+                 "       graph-gqa-cpp [cpu|gpu] <mb> <seq_len> <q_head_num> "
+                 "<kv_head_num> <head_size>\n\n";
+    throw std::invalid_argument("Incorrect input arguments.");
+}
+
+void bench(engine::kind ekind, dnnl_data_type_t dt, const gqa_dims_t &p,
+        double time_limit = 0.) {
+    try {
+        bench_gqa(ekind, static_cast<logical_tensor::data_type>(dt), p,
+                time_limit);
+        get_mem_pool().clear();
+    } catch (dnnl::error &e) {
+        // Catch and report unimplemented cases.
+        if (e.status == dnnl_unimplemented) {
+            std::cout << "unsupported gqa" << std::endl;
+        } else
+            throw;
+    }
+}
+
+void gqa_perf(engine::kind ekind, int argc, char **argv) {
+    // default testing parameters
+    gqa_dims_t params = {2, 384, 16, 2, 64};
```

Comment:
Would be nice to align these values with Tao's PR decreasing examples runtime (unless they already are).

Reply:
Fixed, thanks.